deeplearning/src/main/java/nl/tudelft/bep/deeplearning/clustering/Mapping.java

```diff
@@ -0,0 +1,383 @@
+		for(int i = 0; i < matrix.size(); i++) {
+			for(int j = 0; j < indices.size(); j++) {
+				writer.print(matrix.get(i).get(indices.get(j)));
+				if(j != indices.size()-1) {
+					writer.print(",");
+				}
+			}
+			writer.print("\n");
+		}
+		writer.close();
+	}
+	
+	/**
+	 * Creates a output file based on which layer is representing the pixels of the matrix.
+	 * @param matrix the gene activation matrix
+	 * @param listLayer The layer representing the pixels
+	 * @throws FileNotFoundException
+	 * @throws UnsupportedEncodingException
+	 */
+	public static void writeAvgFile(ArrayList<ArrayList<Double>> matrix, ArrayList<Cluster> listLayer) throws FileNotFoundException, UnsupportedEncodingException {
+		PrintWriter writer = new PrintWriter("sample_dataAVG.dat", "UTF-8");
+		writer.println(listLayer.size());
+		for(int i = 0; i < matrix.size(); i++) {
+			for(int j = 0; j < listLayer.size(); j++) {
+				Double res = avgActivation(matrix, listLayer.get(j), i);
+				writer.print(res);
+				if(j != listLayer.size()-1) {
+					writer.print(",");
+				}
+			}
+			writer.print("\n");
+		}
+		writer.close();	
+	}
+	
+	/**
+	 * Based on the matrix and the pixel at hand, return the average value of the cluster
+	 * @param matrix containing all the gene activation data
+	 * @param c Cluster representing the pixel we want to fill
+	 * @param patient which patient we wish to pick.
+	 * @return
+	 */
+	public static Double avgActivation(ArrayList<ArrayList<Double>> matrix, Cluster c, int patient) {
+		Double res = 0.0;
+		for(Integer i : c.listIndices()) {
+			res += matrix.get(patient).get(i);
+		}
+		return res/c.listIndices().size();
+	}
+	
+	/**
+	 * Given a root cluster this function will give a list of the correlating genes next to each other.
+	 * @param cluster Root cluster of the cluster tree.
+	 * @return List with the points sorted for insertion into the matrix.
+	 */
+	public static ArrayList<Integer> createList(Cluster cluster) {
+		ArrayList<Integer> list = new ArrayList<Integer>();
+
+		if (!cluster.getSet().isEmpty()) {
+			for (Cluster c : cluster.getSet()) {
+				list.addAll(createList(c));
+
+			}
+			return list;
+		} else {
+			list.add(cluster.getID());
+			return list;
+		}
+	}
+
+	/**
+	 * Function which uses the all nearest neighbors algorithm to cluster
+	 * clusters together into a higher level cluster. Clusters can also
+	 * represent points by having an empty set of clusters.
+	 * 
+	 * @param layer1
+	 *            Array of clusters which we would like to cluster
+	 * @return Array containing the newly created higher level clusters.
+	 */
+	public static Cluster[] createClusters(Cluster[] layer1) {
+		// Run nearest neighbor algorithm.
+		NeighborDistance result1 = neighbour(layer1.clone());
+		// ArrayList for keeping track which cluster is already clustered.
+		ArrayList<Boolean> inCluster = new ArrayList<Boolean>(layer1.length);
+		for (int i = 0; i < layer1.length; i++) {
+			inCluster.add(i, false);
```

Comment:
index in add is not used, adding to a list without will also add elements to the end of the list.